python/tvm/autotvm/measure/measure_methods.py

```diff
@@ -0,0 +1,344 @@
+
+    Returns
+    -------
+    res_pack : Array of MeasureResult
+        The list of execution results of measurement.
+    """
+
+    def _fbuild(inp):
+        """ Local build function """
+        with inp.target:
+            s, args = inp.task.instantiate(inp.config)
+            if not inp.config.valid():
+                raise InstantiationError(inp.config.errors)
+            code_hash = getattr(s, 'code_hash', None)
+            if inp.config.code_hash != code_hash:
+                raise HashMismatchError('got {0:s}, expected {1:s}'
+                                        .format(str(inp.config.code_hash), str(code_hash)))
+
+            opts = build_option or {}
+            ctx = context(str(inp.target), 0)
+            if "cuda" in inp.target.keys:
+                # Add cuda verify pass to filter out invalid configs in advance.
+                # This can accelerate the tuning process
+                opts["add_lower_pass"] = \
+                    [(2, cuda_verify_pass(ctx.max_shared_memory_per_block,
+                                          ctx.max_threads_per_block))]
+            with build_config(**opts):
+                func = build(s, args, target_host=inp.task.target_host)
+
+        time_f = func.time_evaluator(
+            func.entry_name, ctx, number=number, repeat=repeat)
+        return time_f, ctx, args
+
+    ret = _measure_generic(_fbuild, input_pack, check_correctness)
+    return ret
+
+
+def cuda_verify_pass(**kwargs):
+    """Verify the invalidity of a cuda kernel
+    This pass will check shared memory size and number of thread per block.
+    """
+    def verify_pass(stmt):
+        valid = ir_pass.VerifyGPUCode(stmt, kwargs)
+        if not valid:
+            raise InstantiationError("invalid cuda kernel")
+        return stmt
+    return verify_pass
+
+
+_cuda_target_arch = None
+@register_func
+def tvm_callback_cuda_compile(code):
+    """use nvcc to generate ptx code for better optimization"""
+    global _cuda_target_arch
+    ptx = nvcc.compile_cuda(code, target="ptx", arch=_cuda_target_arch)
+    return ptx
+
+def set_cuda_target_arch(arch):
+    """set target architecture of nvcc compiler"""
+    global _cuda_target_arch
+    _cuda_target_arch = arch
+
+
+def get_cuda_device_info(device_key, rpc_tracker_addr=None, tmp_dir=None):
+    """get device query result from remote cuda device
+
+    Parameters
+    ----------
+    device_key: str
+        The device key of registered device in RPC tracker.
+    rpc_tracker_addr: Tuple(string, int), optional
+        The address of rpc tracker.
+        If is none, will use environment variable
+    tmp_dir: tvm.contrib.TempDirectory, optional
+        Temporary directory to store cache
+
+    Returns
+    -------
+    max_shared_memory_per_block: int
+    max_threads_per_block: int
+    """
+    if tmp_dir:
+        file_name = tmp_dir.relpath(".autotvm_cuda_device_query_%s.cache" % device_key)
+        find_cache = False
```

Comment:
Cache here seems to be over-designed. Instead,  keep a global scope env, and query it once when we start tuning. Use the tracker addr as key

Reply:
If we want to use cache, there is problems of difference in tracker addr(tune multiple gpus), take a look at tvm.contrib.memoize

Reply:
return a dict